### PR TITLE
Add option to show live cost of 6-cost developments.

### PR DESCRIPTION
**Pull request by jlebar**
<git-pr-chain>

#### Commits in this PR
1. Add live cost display for six-point developments
1. Fix Federation-Capital double-scoring
1. Trying to simplify a bit...
1. simplifying final scoring a bit
1. move logic to client side
1. Handle incremental military and prestige changes
1. More simplifications?
1. Another round, this time including chatgpt pro.
1. More simplifications etc.
1. Tweak two-digit spacing
1. CSS size tweaks
1. Some manual fixes
1. Review comments
1. Rename six_point_dev(elopment) -> six_cost_dev.
1. Coalesce six-point card updates
1. Simplify nomoreVp.
1. Get rid of new game state.
1. Make VP previews visible after choosing card to play.
    
    Before, they'd disappear after a card moved from your hand into the
    tableau.  They'd reappear only after you paid for the card.
1. Client-side fixes.
    
    Previously, after you selected a card to play from your hand into the
    tableau, but before you paid for the card, the card's live VP display
    would go away.  It wouldn't come back even after you cancelled playing
    the card and it went back into your hand.
1. Address review comment.
1. Fix extra notifications
1. Improve animiation code a bit, per review.
1. Actually fix live badges when you play a card
1. Fix animation.
1. Revert "Fix animation."
    
    This reverts commit 1430a08010abe524940508b3bce4f90b751662a3.
1. Fix tableau -> hand animation by using z-index.
1. Don't show a tooltip if we play a card and then leave the mouse hovering where the card *used to* be.
1. Switch to one hex instead of two.
1. Disable font autoscaling on mobile.
1. Hide just-played cards from the live VP scores, so we don't leak info.
1. Tweak location of var-point display on settlements.
    
    Settlement cards' VP hexes are in a sliiiightly different place than
    the hexes on development cards.
1. Address review comments.
1. Add default values.
    
    There seems to have been some update dropping some implicit default values
    thus causing errors during game creation.
1. Fix two live VP bugs
    
    1. The update for the card just played did not work (i.e. playing a development gives the point for Galactic Federation only at the end of the development phase [i.e. after everybody has played] and not directly after [hidden] placement).
    
    2. The scores for the cards in hand vanish after an update. That is it they are correct after a reload, but when updateSixCostDevelopmentVp happens the display changes back to the question mark.

#### [PR chain](https://github.com/jlebar/git-pr-chain)
1. 👉 #5 👈 **YOU ARE HERE**

</git-pr-chain>

### CODE REVIEW

**Review by houseofsuns — COMMENTED** [review]
This is driven by the client side which goes against basically all the other code that is mostly driven by the server side. I'm somewhat unsure whether this is a good idea or whether it would be much better to do it differently.

**Comment by jlebar** [comment]
> After pondering this a bit I think switching from client-side driven to server-side driven would improve the code. Mostly because the code would be less asynchronous that way making things simpler. The second point is then aligning better with the current code base.

That's fine by me but I'm not entirely sure wym.

Are you saying that the client shouldn't call a function to schedule the refresh of the 6-point developments, it should be poked to do so by the server?

**Comment by jlebar** [comment]
I haven't gone over the codex-written code carefully, still just iterating on the UI.  (But I did ask codex to address your comments re server vs client-driven and re showing the value of a card in hand as including the value it would provide by being played.)

Here's my favorite UI so far:

Screenshot 2026-03-12 at 4 39 39 PM

What do you think of this?  It reuses the visual language of the game that "number in a hexagon" means "VPs".

I don't totally love it, but I'm not sure how to improve.

- I tried different colors, but I found it confusing because VPs are always in black-and-white.
- I considered using an outline instead of a solid number, but the number isn't large enough to make that legible, imo. Screenshot 2026-03-12 at 4 46 44 PM
- Outline plus dark gray number.  Also doesn't read super well imo.  We might be able to tweak the colors further but I'm concerned about introducing more visual language into a game where everything already has meaning. Screenshot 2026-03-12 at 4 49 23 PM
 - Put the cost directly inside the hex that currently contains `?`.  The problem is I didn't like covering up the `?` altogether; it sort of makes the UI feel less like a board game.  Also I couldn't figure out a good way to specify that there is an implicit `?` in the points that I'm showing.
 - Something like I had in https://github.com/houseofsuns/raceforthegalaxy/issues/4#issuecomment-4008384296, with maybe a bigger pip maybe in a different location.  I also couldn't make this into something that felt satisfying.  Especially if I had to use e.g. `=2`, because that's rather wide.

**Comment by jlebar** [comment]
> It should then update the player scores accordingly. This needs a new database column to keep track of the share of points attributed to 6-cost devs.  [... This] reduces the load (calculating only once instead of for each player)

FTR I didn't do this.  Felt like a premature optimization, esp because we don't know offhand the cost of the extra work in the database vs the extra work in the PHP?  Either way it feels minimal, probably better to keep the db normalized, i.e. not to store repetitive data?

**Comment by houseofsuns** [comment]
I like the UI with the additional hex. It will encroach a bit on the space for the name of the card. If this is a real problem we can fiddle with that separately, but maybe we are lucky and it's not too bad.

> > It should then update the player scores accordingly. This needs a new database column to keep track of the share of points attributed to 6-cost devs.  [... This] reduces the load (calculating only once instead of for each player)
> 
> FTR I didn't do this. Felt like a premature optimization, esp because we don't know offhand the cost of the extra work in the database vs the extra work in the PHP? Either way it feels minimal, probably better to keep the db normalized, i.e. not to store repetitive data?

This is actually a good point. This moves some business logic to the browser side, but I think that may be the better choice as we can that way stay closer to the current flow of events (thus causing less regressions).

File: raceforthegalaxy.game.php [hunk 1]
```diff
@@ -1106,6 +1116,84 @@ function send_defered_notif($notif_ref)
...
+            'scavengeFromExplore' => true,
+            'scavengerUpdate' => true,
+            'showTableau' => true,
+            'takeover' => true,
+            'updateMilforce' => true,
+            'updatePrestige' => true,
+            'updateScore' => true,
+            'updateSpecializedMilitary' => true,
+        );
+        return isset($relevant_notifications[$notif_type]);
```
Comment (houseofsuns) [inline raceforthegalaxy.game.php]: It would be safer to invert the list, i.e. by default assume that it affects variable vp and only exclude specific notifications. That way a new notification maybe produces more work than necessary, but does not produce a glitch.

File: gamepreferences.json [hunk 2]
```diff
@@ -180,5 +180,18 @@
...
+        "needReload": true,
+        "values": {
+            "1": {
+                "name": "Off"
+            },
+            "2": {
+                "name": "On"
+            }
+        },
+        "default": 2
```
Comment (houseofsuns) [inline gamepreferences.json]: I'm actually still against making this configurable, but not strongly.

File: raceforthegalaxy.game.php [hunk 3]
```diff
@@ -20,6 +20,9 @@ class RaceForTheGalaxy extends Bga\GameFramework\Table
     private $notif_defered_id = 1;
     private $bUpdateCardCount = false;
     private $bUpdateCardCountDefered = false;
+    private $suppressVariableVpNotification = false;
+    private $pendingVariableVpNotification = false;
+    private $variableVpDisplayDisabled = false;
```
Comment (houseofsuns) [inline raceforthegalaxy.game.php]: However if it is configurable it needs to be configurable per player.

File: gamepreferences.json (hunk 2 above)
Comment (jlebar) [reply gamepreferences.json]: I still lean towards making it a configuration option bc in a way it fundamentally changes the "board-game-ness" of the game.  And it necessarily changes the look and feel of the cards.  But I'm fine defaulting it to "on"...

Comment (houseofsuns) [reply gamepreferences.json]: I can work with a default "on". This however means that it will always be a frontend mechanism and never touch the points recorded in the database (which is actually a good idea).

File: cards.inc.php [hunk 4]
```diff
@@ -9065,7 +9065,7 @@
         'phases' => array( 2),
         'points' => 3,
         'name' => clienttranslate("Galactic Status"),
-        'description' => clienttranslate("First to place a 6-cost &lt;?&gt; development.")
+        'description' => clienttranslate("First to place a six-point &lt;?&gt; development.")
```
Comment (houseofsuns) [inline cards.inc.php]: I think this is worse than before. They actually cost 6 and only in rare cases score 6 points. This also applies to the power change above.

Also a translated string should only be changed if really necessary as this will produce churn for the translators (and many minor errors can also fixed during translation).

PS: I just saw that there are a lot more similar changes. I don't think they are worth the hassle / beneficial at all.

**Review by houseofsuns — CHANGES_REQUESTED** [review]

File: cards.inc.php (hunk 4 above)
Comment (jlebar) [reply cards.inc.php]: Completely agree.  I haven't read this PR carefully yet (just have it up as a draft), was focusing on functionality and getting the UI right.

Comment (houseofsuns) [reply cards.inc.php]: A sorry, I mixed up what's ready and what's not.

Comment (jlebar) [reply cards.inc.php]: (Sorry it's been slow, reviewing this AI-generated code is such a slog!)

File: raceforthegalaxy.game.php (hunk 1 above)
Comment (jlebar) [reply raceforthegalaxy.game.php]: I just went with recalculating it all the time because compute is cheap, but happy to add a blacklist if you want!

**Comment by jlebar** [comment]
Okay, my eyes are glazing over from having read this PR a bunch of times.  I *think* it's decent?  Would you like to have a look?

Note that there's a fair amount of moved code here.  I found that `git diff --color-moved=dimmed-zebra` worked somewhat decently.

File: cards.inc.php (hunk 4 above)
Comment (jlebar) [reply cards.inc.php]: Should be better now.

File: cards.inc.php
Comment (houseofsuns) [inline cards.inc.php:9234]: Maybe add a comment that this now are variable point cards (contrary to the name).

File: raceforthegalaxy.css [hunk 5]
```diff
@@ -304,6 +304,14 @@ html.new_design {
     transform-origin: top right;
     scale: 75%;
 }
+.card_size_tiny .six_point_development_live_value {
```
Comment (houseofsuns) [inline raceforthegalaxy.css]: This currently contains lots of "six point" instead of "six cost" denominations. I think this is not a big problem as it can be cleaned up at any point in time by some advanced search and replace.

File: raceforthegalaxy.game.php [hunk 6]
```diff
@@ -1108,6 +1107,28 @@ function send_defered_notif($notif_ref)
...
+    // function also recalculates the live six-point development scores and
+    // sends them to all players as well, as a separate notification.
+    //
+    // (We could instead try to keep a list of which events should trigger us to
+    // recalculate the live six-point development scores, but that would be
+    // fragile, whereas this does some extra work but is always correct.)
+    function notifyAllPlayers($notif_type, $notif_log = '', $notif_args = array())
+    {
+        parent::notifyAllPlayers($notif_type, $notif_log, $notif_args);
+        $this->emitLiveSixPointDevelopmentState($this->buildLiveSixPointDevelopmentDisplayState());
```
Comment (houseofsuns) [inline raceforthegalaxy.game.php]: This should probably be a "maybe emit" which includes the caching I previously mentioned (only creating a new message if there is some actual change) to not blow up the history size to much.

File: raceforthegalaxy.game.php [hunk 7]
```diff
@@ -3416,18 +3479,18 @@ function getPossibleConsumptionCards($player_id_filter = null, $init_phase = fal
         return $result;
     }
 
-    function cardsToSixDevelopmentsScore($cards, $dev_to_players, $player_infos, $oort_player = null, $oort_value = null)
+    function cardsToSixDevelopmentsScore($cards, $dev_to_players, $player_infos, $oort_player = null, $oort_value = null, $change_oort_type = true)
```
Comment (houseofsuns) [inline raceforthegalaxy.game.php:3503]: This is a bit unsatisfying and I'm a bit unsure what the best way forward is here. If we do the whole dance with testing all Oort variants we will produce values that are not valid as counted on the board. But if we don't do this there may be a change at the end of the game.
However I think the latter is much less confusing and I would vote for suppressing the Oort variants for the live counting.

File: raceforthegalaxy.js [hunk 8]
```diff
@@ -838,6 +840,9 @@ define([
                 dojo.query('.prestigeleadercount').forEach(dojo.hitch(this, function(node) {
                     node.innerHTML = this.gamedatas.prestigeleadercount;
                 }));
+                this.currentStateName = (this.gamedatas.gamestate || {}).name || null;
```
Comment (houseofsuns) [inline raceforthegalaxy.js]: This feels very defensive. Any idea why this is necessary? (Other places use `this.gamedatas.gamestate.name` directly.)

File: states.inc.php [hunk 9]
```diff
@@ -653,7 +653,15 @@
...
-        "transitions" => array("endGame" => 98, "nextRound" => 10)
+        "transitions" => array("finalScoring" => 97, "nextRound" => 10)
+   ),
+    97 => array(
+        "name" => "finalScoring",
+        "description" => '',
+        "descriptionmyturn" => '',
+        "action" => "stFinalScoring",
+        "type" => "game",
+        "transitions" => array("" => 98)
```
Comment (houseofsuns) [inline states.inc.php]: I don't think this will trigger any surprises during roll-out, but we should think carefully here to not break any running games.
PS: I'm not entirely sure why we add this state, but I presume that otherwise there would be some display artifacts.

**Review by houseofsuns — COMMENTED** [review]
I looked at the overall structure and I'm pleased by what I see. I did not yet have the time to dig into all the details. I also did a test run on the Studio and liked what I saw. This is shaping up nicely.

File: raceforthegalaxy.js (hunk 8 above)
Comment (jlebar) [reply raceforthegalaxy.js]: Fixed, ty

File: raceforthegalaxy.game.php (hunk 7 above)
Comment (jlebar) [reply raceforthegalaxy.game.php:3503]: Agreed, done.

File: raceforthegalaxy.css (hunk 5 above)
Comment (jlebar) [reply raceforthegalaxy.css]: Yes, happy to do it whenever, as part of this patch or another one.

File: cards.inc.php
Comment (jlebar) [reply cards.inc.php:9234]: Done

File: states.inc.php (hunk 9 above)
Comment (jlebar) [reply states.inc.php]: Yes I think so.  When we enter endgame scoring we need an additional state to say "we're scoring cards but haven't finished yet".  But because this state comes and goes without any user intervention, I and Codex agree that I *think* it can't break in-progress games.

File: raceforthegalaxy.game.php (hunk 6 above)
Comment (jlebar) [reply raceforthegalaxy.game.php]: For my info, how does the blown-up history manifest?

I'm happy to do this if you think it's important, but it's kind of ugly.  AFAICT we have to make a new DB table to remember what we last sent to each user.  And then we have to handle the possibility that the table doesn't exist...  And I'm also unclear what happens if a client misses one of these notifications due to a network flake or whatever.  Will we send them another one when they reconnect?  How does it work, exactly?

As a weaker version we could remember the game state as a PHP member variable, but then I don't know how many duplicates we'd be able to filter out.

**Comment by jlebar** [comment]
Pushed a new version, please have a look!

File: raceforthegalaxy.css (hunk 5 above)
Comment (houseofsuns) [reply raceforthegalaxy.css]: I'd say: go for it. The earlier this is rectified the higher the chance to find any remaining issues.

File: raceforthegalaxy.game.php (hunk 6 above)
Comment (houseofsuns) [reply raceforthegalaxy.game.php]: I thought of a member variable containing an associative array mapping player ids to last private state sent with some default value (maybe zero) mapping to the last global state sent.
I'm not sure how this works on the BGA server side; I guess they serialize PHP state on inactivity. However if this at some point flakes we may sent out some additional notifications which is not a problem.
If the client has problems they should probably reload getting a fresh state and otherwise they will catch up with the next actual change.
For the context: the replay feature works by recording all user inputs and notifications and then rerunning the client side code. I don't know the exact details as this is a platform feature that needs basically no thought (besides some basic rules like not triggering actions from notifications). Most notifications are rather small (like updating the score contains an integer of points per player). If we now amplify all of these notifications with a probably heavier one this will drastically increase the storage necessary to save a replay. As the complexity of the caching check should be rather low I think this is a worthwhile optimization (if this needed new DB resources that tradeoff would look quite different).

File: states.inc.php (hunk 9 above)
Comment (houseofsuns) [reply states.inc.php]: What exactly would happen if we did not take care and insert an extra state and hide the counts during final scoring?
This is probably obvious once you see it, but currently I'm stumped.

File: raceforthegalaxy.game.php [hunk 10]
```diff
@@ -2501,23 +2538,16 @@ function onEndAjaxAction()
...
-        } else {
-            $current_milforce = self::getUniqueValueFromDB("SELECT player_xeno_milforce FROM player WHERE player_id='$player_id'");
-        }
-
         $new_milforce = 0;
         $military_world_count = 0;
         $chromosome_world_count = 0;
         $civil_world_count = 0;
         $imperium_count = 0;
+        $has_imperium = false;
```
Comment (houseofsuns) [inline raceforthegalaxy.game.php]: $has_imperium is redundant, just check whether $imperium_count is non-zero.

File: raceforthegalaxy.game.php [hunk 11]
```diff
@@ -3416,45 +3479,48 @@ function getPossibleConsumptionCards($player_id_filter = null, $init_phase = fal
...
-                    3 => $this->cardsToSixDevelopmentsScore($cards, $dev_to_players, $player_infos, $oort_player, 3),
-                    4 => $this->cardsToSixDevelopmentsScore($cards, $dev_to_players, $player_infos, $oort_player, 4)
-               );
+        if ($oort_player !== null && $oort_value === null) {
+            // We call cardsToSixDevelopmentsScore FOUR times with all possible values, and get the best configuration for Oort
+            $good_to_result = array(
+                1 => $this->cardsToSixDevelopmentsScore($cards, $dev_to_players, $player_infos, $oort_player, 1, $change_oort_type),
+                2 => $this->cardsToSixDevelopmentsScore($cards, $dev_to_players, $player_infos, $oort_player, 2, $change_oort_type),
+                3 => $this->cardsToSixDevelopmentsScore($cards, $dev_to_players, $player_infos, $oort_player, 3, $change_oort_type),
+                4 => $this->cardsToSixDevelopmentsScore($cards, $dev_to_players, $player_infos, $oort_player, 4, $change_oort_type)
```
Comment (houseofsuns) [inline raceforthegalaxy.game.php]: These calls should never change the Oort type as this is done below after calculating the best type. So I think we should pass "false" here.

File: raceforthegalaxy.js [hunk 12]
```diff
@@ -4773,6 +4925,8 @@ define([
 
                 dojo.subscribe('updateCardCount', this, "notif_updateCardCount");
                 dojo.subscribe('updateScore', this, "notif_updateScore");
+                dojo.subscribe('updateSixPointDevelopmentVp', this, "notif_updateSixPointDevelopmentVp");
+                dojo.subscribe('nomoreVp', this, "notif_nomoreVp");
```
Comment (houseofsuns) [inline raceforthegalaxy.js]: I'm a bit unsure about the details here. Previously the notification was just a string and got displayed as popup toast. Now we hide the live point display. Does this change suppress the popup toast or does it still show up?

**Review by houseofsuns — COMMENTED** [review]
I took a closer look and the details look really promising. Just some minor remarks here.

File: raceforthegalaxy.game.php (hunk 6 above)
Comment (jlebar) [reply raceforthegalaxy.game.php]: If we can do it with member variables then I agree we should, that's easy.  But I don't think it works.

1. If we could store state in PHP member variables, then why do we use this get/setGameStateValue mechanism when instead we could simply store member variables for all this other stuff?  Indeed why have a SQL database at all?

2. The docs are pretty clear I think?  https://en.doc.boardgamearena.com/Main_game_logic:_Game.php

> Your php class instance won't be in memory between two callbacks, every time client send a request a new class will be created, constructor will be called and eventually your callback function.

File: raceforthegalaxy.css (hunk 5 above)
Comment (jlebar) [reply raceforthegalaxy.css]: Done!

File: raceforthegalaxy.game.php (hunk 6 above)
Comment (jlebar) [reply raceforthegalaxy.game.php]: WRT saving the replays:

> The bundle of notifications sent at the end of an action is considered a single "move" and the table's move counter increases.

https://en.doc.boardgamearena.com/Main_game_logic:_Game.php

So the fact that we're sending a larger *number* of notifications shouldn't affect the replay experience for users.  But I agree the replay *storage* does increase dramatically.  It looks like the limit is 64MB per game, which surely we're not going to get anywhere near even if we're wasteful like this?  Still I hear the concern...

Chat suggests that we can coalesce these updates so they only get sent once per action, instead of once per notification.  That would be a big improvement.  I'll try it out!

File: states.inc.php (hunk 9 above)
Comment (jlebar) [reply states.inc.php]: I think the change to coalesce the notifications so they are sent only once per action may also let us get rid of this new game state, let me see...

File: raceforthegalaxy.js (hunk 12 above)
Comment (jlebar) [reply raceforthegalaxy.js]: I think this isn't necessary anymore, removed!

For completeness here's codex's answer to your original question.

----

It should still show up.

nomoreVp is still sent with a non-empty log string from PHP, for example "No more victory point chip: end of game" (raceforthegalaxy.game.php (line 10742)). On the client, dojo.subscribe('nomoreVp', ...) just attaches an extra handler; it does not replace or suppress BGA’s normal notification logging (raceforthegalaxy.js (line 4928), raceforthegalaxy.js (line 5564)). The bundled BGA docs also describe notifications as carrying both log and args, with the handler simply receiving that notification object, not taking over log display (Game_interface_logic__Game.js.wiki).

So the effect of this change is additive: the popup/log message should still appear, and the handler also hides the live point display. If you wanted to suppress the popup, you’d need to change the notification/log behavior itself, not just add a JS subscriber.

File: raceforthegalaxy.game.php (hunk 6 above)
Comment (jlebar) [reply raceforthegalaxy.game.php]: I've changed the code so that we don't send any unnecessary notifications.

- At the beginning of each action, we compute and save the six-cost state.
- At the end of each action, we send the new state if and only if it has changed.

Much cleaner!

File: states.inc.php (hunk 9 above)
Comment (jlebar) [reply states.inc.php]: I got rid of it.  I think the downside is that during final game scoring there is an *extremely brief* period of time when the RHS VP counter will be wrong.

AIUI we send updateScore notifications during final scoring, and those increment the VPs on the RHS.  If you have the live six-cost development display enabled, the six-cost devs will be included in that score, so we're double-counting.  Then the game actually ends, and we stop counting the six-cost devs in there, and the score is correct.

The point of the new state was to enter the "stop counting live six-point devs in the RHS score" state before the game actually ends.  But since that state exists only for as long as it takes the client to process all the updateScore notifications, I don't think it matters / is user-visible.

**Comment by jlebar** [comment]
I've updated the code, please have a look!

File: raceforthegalaxy.game.php [hunk 13]
```diff
@@ -2499,20 +2525,38 @@ function onEndAjaxAction()
...
-    {
-        $cards = $this->cards->getCardsInLocation('tableau', $player_id);
+        // Update the live six-cost development scores if they have changed.
+        $initial_state = $this->initialSixCostDevPointsState;
+        $this->initialSixCostDevPointsState = null;
+        $state = $this->buildLiveSixCostDevelopmentDisplayState();
+        // initial_state could be null if we failed to capture it at the
+        // beginning of the action.  This is a bug, but we can defensively
+        // handle it by sending the notification anyway.
+        if ($initial_state === null || $state != $initial_state) {
```
Comment (houseofsuns) [inline raceforthegalaxy.game.php]: `$state != $initial_state` should cover all cases. If `$initial_state` is null it differs from the array `$state`.

File: raceforthegalaxy.game.php (hunk 6 above)
Comment (houseofsuns) [reply raceforthegalaxy.game.php]: You are totally correct. I confused the client and server side mechanics. *blush*
The new solution really convinces me!

File: states.inc.php (hunk 9 above)
Comment (houseofsuns) [reply states.inc.php]: I tried this and I could not see a wrong number; it's gone too fast.

File: raceforthegalaxy.game.php (hunk 6 above)
Comment (houseofsuns) [reply raceforthegalaxy.game.php]: I tried this on the Studio and it mostly worked. However there seem to be some framework callbacks that cause spurious notifications. In my case lots of calls to "wakeup.html" answered with "yourturnack" and "updateSixCostDevelopmentVp" notifications.

**Review by houseofsuns — COMMENTED** [review]
We are approaching a good state. I tested this on the Studio and the only thing that I noticed is that the live counting hex vanished when playing a card until it is revealed to all other players. It also stayed gone if I cancelled a play and the card returned to my hand. This is somewhat irritating but would not count as show-stopper in my eyes.

**Comment by jlebar** [comment]
I've updated the code to fix all your comments other than the spurious notifications issue discussed below!

> I tested this on the Studio and the only thing that I noticed is that the live counting hex vanished when playing a card until it is revealed to all other players. It also stayed gone if I cancelled a play and the card returned to my hand. 

Good catch, fixed.

> However there seem to be some framework callbacks that cause spurious notifications. In my case lots of calls to "wakeup.html" answered with "yourturnack" and "updateSixCostDevelopmentVp" notifications.

Hm, I haven't been able to reproduce this.  I started a two-player game in studio and opened the Chrome developer console in the "network" tab in both games.

I see requests to wakeup.html, e.g. `https://studio.boardgamearena.com/1/raceforthegalaxy/raceforthegalaxy/wakeup.html?myturnack=true&table=872450&noerrortracking=true&testuser=2431132&dojo.preventCache=1775759437744`

But the response is always short, something like `{"status":1,"data":{"valid":1,"data":null}}`.

I must be looking at this wrong somehow, because I actually don't see all of the "update six-cost scores" messages that I expect to see based on what's happening in the game.

**Comment by houseofsuns** [comment]
I used the log built into Studio instead of the developer console. You find it when scrolling down below the buttons to save and reload previous game states (see image for an example).
studio-inout

File: raceforthegalaxy.js [hunk 14]
```diff
@@ -954,6 +954,13 @@ define([
                 var card_type = this.gamedatas.card_types[card_type_id];
                 return card_type && typeof card_type.sixdev_scoring != 'undefined';
             },
+            returnCardToHand: function(card) {
+                this.playerHand.addToStockWithId(card.type, card.id, 'card_wrapper_' + card.id);
+            },
```
Comment (houseofsuns) [inline raceforthegalaxy.js:987]: These look asymmetric. I would add the `dojo.destroy($('card_wrapper_' + this.nextCardToPlay.id));` to `returnCardToHand`.
Also I'm a bit confused that `dojo.destroy('player_hand_item_' + notif.args.card.id);` does not appear in `moveHandCardToTableau`.

**Comment by jlebar** [comment]
Thanks, I reproduced and fixed the spurious six-cost messages.  It's a little dicey because it relies on the fact that any action which affects the game state must go through actions.php.  Whereas before we were being defensive and saying that anything which *doesn't* go through actions.php is considered to have a dirty six-cost dev state.  But it seems to work fine this way.

Somehow I un-fixed `the live counting hex vanished when playing a card` though.  I'll let you know when I have a fix for that...

**Comment by jlebar** [comment]
> Somehow I un-fixed the live counting hex vanished when playing a card though. I'll let you know when I have a fix for that...

Fixed!  PHAL!

File: raceforthegalaxy.js [hunk 15]
```diff
@@ -954,8 +954,41 @@ define([
...
             },
+            getReturnCardToHandAnimationDuration: function() {
+                return this.bgaAnimationsActive() ? 1000 : 0;
+            },
+            // For some reason, the animation for moving a card from a player's
+            // tableau back into their hand is janky: The card itself isn't
+            // visible for the first ~half of the animation.  This presents a
+            // problem for the six-cost development live VP badges; if we don't
+            // do anything, then the badge itself flies back into the hand while
+            // the card itself is invisible.  It looks weird!
```
Comment (houseofsuns) [inline raceforthegalaxy.js]: This feels strange. It seems that the card vanishes below the tableau background. I could experimentally fix this by adding `z-index: -3;` to the `.whiteblock` css class. That sounds way simpler. Or am I missing something here?

Comment (houseofsuns) [reply raceforthegalaxy.js]: Argh, this breaks onClick events for the tableau. E.g. Consume doesn't work anymore because you can't click on goods or worlds.

Comment (jlebar) [reply raceforthegalaxy.js]: Good call figuring out that it's a z-index issue though.

We could temporarily increase the z-index of the card while it's animating?  lol

Comment (houseofsuns) [reply raceforthegalaxy.js]: Not as simple as my first attempt, but the machinery is already there to modify the animating card, so why not use it?

**Comment by houseofsuns** [comment]
> Thanks, I reproduced and fixed the spurious six-cost messages. It's a little dicey because it relies on the fact that any action which affects the game state must go through actions.php. Whereas before we were being defensive and saying that anything which _doesn't_ go through actions.php is considered to have a dirty six-cost dev state. But it seems to work fine this way.

Sounds reasonable. Let's do this.

File: raceforthegalaxy.js (hunk 15 above)
Comment (jlebar) [reply raceforthegalaxy.js]: Done and uploaded, please have a look.

I also fixed an issue I noticed with tooltips (see commit message).  It may have been a pre-existing issue, I'm not sure.

**Comment by houseofsuns** [comment]
I took a quick look and this really looks good. As far as I'm aware there are no open issues remaining. So as I see it we are now in stabilization and should test any remaining edge cases.
As this is mostly a client side thing it is luckily rather low risk.

**Comment by jlebar** [comment]
> So as I see it we are now in stabilization and should test any remaining edge cases.

I had codex run some tests (with the Chrome MCP server it can use a browser) and it didn't find any problems, but I wouldn't put much weight on that.  I have a board game party on Sunday and will try to get some people there to playtest it, maybe they'll catch things that we missed.

**Comment by jlebar** [comment]
I playtested this with a friend who has played a lot of rftg using the bespoke Android app (ie not bga).

I found one legit bug, which is that the numbers don't render well on my phone.

But his main feedback was that he didn't love the "two overlapping hexes" idea we landed on.  Indeed in his Android app variable point cards' points are rendered inline with the points in a slightly different color.  See attached.

I'll definitely fix the mobile rendering.  Wdyt about the single vs two hex approach?
IMG_0256

**Comment by houseofsuns** [comment]
I like the one hex with different color. We added the second hex to not obscure that the VPs are variable. Using a color can convey the same meaning with less visual clutter, so I'm all for it.
However I'm not a big fan of the ochre colour chosen in the screenshot. Maybe something more like teal? (But that's just bike-shedding.)

**Comment by jlebar** [comment]
Tweaked the CSS as discussed.  Also fixed the display on mobile.  [c8e8b53](https://github.com/houseofsuns/raceforthegalaxy/pull/5/commits/c8e8b5358a783701586e0cd5fd405185bc601b9c) fixes the display of the live VP points on cards, and also fixes a preexisting bug where card titles would be displayed too big (for certain card sizes).

**Comment by jlebar** [comment]
I tried a few other colors but stayed with the brownish color.  I felt like brighter colors (including the teals I tried) often clashed with the cards themselves.  Brown is still not great because brown has meaning in RFTG, but I didn't love gray.

Also this brownish color will be familiar to people who have played the official rftg app, maybe that's a bonus?

**Comment by houseofsuns** [comment]
I like the new design. Why didn't we think of this in the beginning?
It's a tiny bit subtle, i.e. it's possible to miss the color hint, but then you can view the tooltip to clarify.

Two more issues. First the minor one: On my mobile Federation Capital showed a small bit of the original icon (see screenshot).
Screenshot_20260417-094839~2

The bigger issue is the info-leak: If player A places a 6-cost dev or a world that gives points via an already placed 6-cost-dev then the score is immediately updated even for player B who does not yet see the card that gives the points. The game already has some info-leaks and I would like to avoid introducing another one (and one that is so in your face; most other info-leaks require reloading).
The solution is probably to in general exclude cards in the `player_just_played` slot and include for each player the individual card there for the private information part.

**Comment by jlebar** [comment]
Ready for another review.

> On my mobile Federation Capital showed a small bit of the original icon (see screenshot).

Good catch.  Verified that the scoring hex on settlements is in a sliiiightly different place than it is for developments.

Elsewhere in this PR, I've endeavored to use "six-cost development" to refer to all variable-point cards.  But because we need different CSS rules for developments and settlements, I decided to rename the CSS classes to "variable_point_card", "variable_point_development", and "variable_point_settlement".

> The bigger issue is the info-leak

Another good catch, and fixed.

File: raceforthegalaxy.css [hunk 16]
```diff
@@ -304,6 +312,19 @@ html.new_design {
...
+.card_size_tiny .variable_point_card {
+    top: 12px;
+    width: 12px;
+    height: 13px;
+    font-size: 10px;
+}
+.card_size_tiny .variable_point_development_live_value {
+    left: 25px;
+}
+.card_size_tiny .variable_point_settlement_live_value {
```
Comment (houseofsuns) [inline raceforthegalaxy.css]: The term "settlement" should be "world" to align with normal usage.

File: raceforthegalaxy.js [hunk 17]
```diff
@@ -5388,6 +5615,29 @@ define([
...
+                };
+                if (notif.args._private) {
+                    if (notif.args._private.card_scores) {
+                        state.private_card_scores = notif.args._private.card_scores;
+                    } else if (notif.args._private[this.player_id] && notif.args._private[this.player_id].card_scores) {
+                        state.private_card_scores = notif.args._private[this.player_id].card_scores;
+                    }
+                    if (notif.args._private.player_totals) {
+                        state.player_totals = notif.args._private.player_totals;
+                    } else if (notif.args._private[this.player_id] && notif.args._private[this.player_id].player_totals) {
```
Comment (houseofsuns) [inline raceforthegalaxy.js]: Are both `notif.args._private.player_totals` and `notif.args._private[this.player_id].player_totals` used? It feels like only one of these branches should be necessary.

File: raceforthegalaxy.game.php [hunk 18]
```diff
@@ -4016,9 +4025,8 @@ private function loadSixCostDevelopmentPlayerInfos()
         return $player_infos;
     }
 
-    private function getSixCostDevelopmentScoringContext()
+    private function getSixCostDevelopmentScoringContext($cards, $recompute_military)
```
Comment (houseofsuns) [inline raceforthegalaxy.game.php]: The `recompute_military` parameter feels superfluous. As far as I can see it's always true except for the final scoring, but there it should not make a difference and is a rather minor optimization. So it seems to me that this adds unnecessary complexity or am I missing something?

**Review by houseofsuns — APPROVED** [review]
Just some suggestions. I just took a look at the code and did not get around to test how it plays.

File: raceforthegalaxy.js (hunk 17 above)
Comment (jlebar) [reply raceforthegalaxy.js]: agreed, thanks, fixed

File: raceforthegalaxy.game.php (hunk 18 above)
Comment (jlebar) [reply raceforthegalaxy.game.php]: agreed, fixed

**Comment by jlebar** [comment]
Addressed all your comments!
